src/Disp/BuildingsUpgrades/Buildings.js

```diff
@@ -63,47 +64,64 @@ export default function UpdateBuildings() {
-  // (grid rows are 1-based indexing, and row 1 is the bulk buy/sell options)
-  // This regulates sorting of buildings
-  if (Game.buyMode === 1 && CMOptions.SortBuildings) {
-    let arr;
-    if (CMOptions.SortBuildings === 1) {
-      arr = Object.keys(CacheObjects1).map(k => {
-        const o = {};
-        o.name = k;
-        o.pp = CacheObjects1[k].pp;
-        o.color = CacheObjects1[k].color;
-        return o;
-      });
-    } else if (CMOptions.SortBuildings === 2) {
-      arr = Object.keys(target).map(k => {
-        const o = {};
-        o.name = k;
-        o.pp = target[k].pp;
-        o.color = target[k].color;
-        return o;
-      });
-    }
+  // Build array of pointers and sort according to the user's configured sort option.
+  // This regulates sorting of buildings.
+  let arr;
+  if (Game.buyMode !== 1 || !CMOptions.SortBuildings) {
+    arr = Object.keys(CacheObjects1).map(k => {
+      const o = {};
+      o.name = k;
+      o.id = Game.Objects[k].id;
+      return o;
+    });
+    // Sort using default order.
+    arr.sort((a, b) => a.id - b.id);
+  } else if (CMOptions.SortBuildings === 1) {
+    arr = Object.keys(CacheObjects1).map(k => {
+      const o = {};
+      o.name = k;
+      o.pp = CacheObjects1[k].pp;
+      o.color = CacheObjects1[k].color;
+      return o;
+    });
     // Sort by pp colour group, then by pp.
     arr.sort((a, b) =>
       ColoursOrdering.indexOf(a.color) === ColoursOrdering.indexOf(b.color)
         ? a.pp - b.pp
         : ColoursOrdering.indexOf(a.color) - ColoursOrdering.indexOf(b.color)
     );
-    for (let x = 0; x < arr.length; x++) {
-      Game.Objects[arr[x].name].l.style.gridRow = `${x + 2}/${x + 2}`;
-    }
-  } else {
-    const arr = Object.keys(CacheObjects1).map(k => {
+  } else if (CMOptions.SortBuildings === 2) {
+    arr = Object.keys(target).map(k => {
       const o = {};
       o.name = k;
-      o.id = Game.Objects[k].id;
+      o.pp = target[k].pp;
+      o.color = target[k].color;
       return o;
     });
-    arr.sort((a, b) => a.id - b.id);
-    for (let x = 0; x < arr.length; x++) {
-      Game.Objects[arr[x].name].l.style.gridRow = `${x + 2}/${x + 2}`;
-    }
+    // Sort by pp colour group, then by pp.
+    arr.sort((a, b) =>
+      ColoursOrdering.indexOf(a.color) === ColoursOrdering.indexOf(b.color)
+        ? a.pp - b.pp
+        : ColoursOrdering.indexOf(a.color) - ColoursOrdering.indexOf(b.color)
+    );
+  } else if (CMOptions.SortBuildings === 3) {
+    arr = Object.keys(CacheObjectsNextAchievement).map(k => {
+      const o = {};
+      o.name = k;
+      o.amountUntilNext = CacheObjectsNextAchievement[k].AmountNeeded;
+      o.priceUntilNext = CacheObjectsNextAchievement[k].price;
+      return o;
+    });
+    // Sort by price until next achievement. Buildings that aren't close to an achievement are always last.
+    arr.sort((a, b) =>
+      (a.amountUntilNext !== 101 ? a.priceUntilNext : Infinity) -
```

Comment:
I'm wondering whether using Infinity here could create problems when the `CacheObjectsNextAchievement` or `GameObjects` (on which it builds) are no longer in order. Especially since Orteil is known to create incompatible changes.
I would prefer if we would sort on `id` if `amountUntilNext` is 101.

Reply:
It's true, we should account for the case where the source data is out of order.

The fix you proposed would not quite work, though, because then the sort logic would be comparing IDs against priceUntilNext. Since sort() is guaranteed to be a stable sort, we can 1) sort by ID first, then 2) do this sort as-is. The array is small, so doing two passes should not impact performance.

Reply:
Perfect fix! Thanks!